src/hotspot/share/opto/library_call.cpp

```diff
@@ -4070,6 +4070,20 @@ bool LibraryCallKit::inline_fp_conversions(vmIntrinsics::ID id) {
   return true;
 }
 
+//----------------------has_wide_mem-------------------------
+bool LibraryCallKit::has_wide_mem(Node* addr, Node* base) {
+  const TypeAryPtr* addr_t = _gvn.type(addr)->isa_aryptr();
+  const Type*       base_t = _gvn.type(base);
+
+  bool in_native = (base_t == TypePtr::NULL_PTR);
+  bool in_heap   = !TypePtr::NULL_PTR->higher_equal(base_t);
+  bool is_mixed  = !in_heap && !in_native;
+
+  bool is_prim_array = (addr_t != NULL) && (addr_t->elem() != Type::BOTTOM);
```

Comment:
What does `is_prim_array` stand for? I read this as "is primitive array" which does not make sense :)

Reply:
Yes, it stands for "primitive array". 
Unsafe.copyMemory() validates inputs using Unsafe.checkPrimitiveArray() to ensure they are primitive arrays for on-heap accesses. I refactored `has_wide_mem` a bit to make it a bit clearer what's going on.